{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "#Students: Elisei Smirnov, Amine Trabelsi, Liana Mardanova\n",
        "## Group: DS-01\n",
        "## Task description: [Link](https://www.kaggle.com/competitions/santa-workshop-tour-2019/data)\n",
        "\n"
      ],
      "metadata": {
        "id": "SQlXBpvvJX9D"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "EuA6Ljcjqori"
      },
      "outputs": [],
      "source": [
        "# Import necessary libraries\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import random\n",
        "import copy\n",
        "from typing import List, Tuple, Dict, Any\n",
        "\n",
        "# Read the family data into a DataFrame, using 'family_id' as the index\n",
        "data_file_path = 'family_data.csv'\n",
        "data = pd.read_csv(data_file_path, index_col='family_id')\n",
        "\n",
        "# Read the sample submission data into a DataFrame, using 'family_id' as the index\n",
        "sample_submission_file_path = 'sample_submission.csv'\n",
        "sample_submission = pd.read_csv(sample_submission_file_path, index_col='family_id')\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F8Frg4Wlcqe0"
      },
      "source": [
        "### Given information"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3TYm5pmcq0Gn"
      },
      "outputs": [],
      "source": [
        "data.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "MCcKJ1NCrA5Y"
      },
      "outputs": [],
      "source": [
        "sample_submission.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4awBY-1Bc3zu"
      },
      "source": [
        "### Penalty cost for Santa\n",
        "\n",
        "> The total number of people attending the workshop each day must be between $125$ - $300$; if even one day is outside these occupancy constraints, the submission will error and will not be scored.\n",
        "\n",
        "> Santa provides consolation gifts to families. These sum up per family, and the total represents the **$preference:cost$**:\n",
        "- `choice_0`: no consolation gifts\n",
        "- `choice_1`: \\$50\n",
        "- `choice_2`: \\\\$50, \\$9 for each family member\n",
        "- `choice_3`: \\\\$100, \\$9 for each family member\n",
        "- `choice_4`: \\\\$200, \\$9 for each family member\n",
        "- `choice_5`: \\\\$200, \\$18 for each family member\n",
        "- `choice_6`: \\\\$300, \\$18 for each family member\n",
        "- `choice_7`: \\\\$300, \\$36 for each family member\n",
        "- `choice_8`: \\\\$400, \\$36 for each family member\n",
        "- `choice_9`: \\\\$500, \\\\$36 for each family member, \\\\$199 for each family member\n",
        "- `otherwise`: \\\\$500, \\\\$36 for each family member, \\$398 for each family member.\n",
        "\n",
        "> $$accoutning:penalty = \\sum_{d=100}^1\\frac{N_d-125}{400}N_d^{\\frac{1}{2}+\\frac{|N_d-N_{d+1}|}{50}}$$\n",
        "> where $N_d$ is the occupancy of the current day, and $N_{d+1}$ is the occupancy of the previous day (since we're counting backwards from Christmas! For the initial condition of $d=100$, $N_{101}=N_{100}$.\n",
        "\n",
        "> **Finally $score=preference:cost+accounting:penalty$**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JqqRk4bqjL4U"
      },
      "source": [
        "### Genetic Algorithm Implementation"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "\"\"\"\n",
        "Some parts of tge code based on already provided by comptetition starter notebook:\n",
        "https://www.kaggle.com/code/inversion/santa-s-2019-starter-notebook\n",
        "\n",
        "\"\"\"\n",
        "\n",
        "class GeneticScheduler:\n",
        "    \"\"\"\n",
        "    This class reads family preference data, generates an initial population, and iteratively\n",
        "    evolves this population to find an optimal schedule that minimizes fitness function. This\n",
        "    function is based on the costs and constraints, provided in the description of Santa's\n",
        "    Workshop Tour 2019 problem.\n",
        "\n",
        "    This class encapsulates the logic for scheduling families based on their preferences\n",
        "    and constraints using a genetic algorithm. It includes methods for initializing the\n",
        "    population, performing crossover and mutation operations, selecting parents, and\n",
        "    evaluating fitness.\n",
        "\n",
        "    Attributes:\n",
        "    - file_location (str): File location of the family preference data.\n",
        "    - answer_location (str): File location to save the best solution.\n",
        "    - num_days (int): Number of days Santa's workshop is open.\n",
        "    - family_num (int): Number of families.\n",
        "    - MIN_PEOPLE (int): Minimum constraint of people in a workshop per day.\n",
        "    - MAX_PEOPLE (int): Maximum constraint of people in a workshop per day.\n",
        "    - preferences_num (int): Number of preferences each family provides.\n",
        "    - population_size (int): Size of the population in each generation.\n",
        "    - mutation_rate (float): Probability of type of mutation for each chromosome.\n",
        "    - crossover_rate (float): Probability of crossover between two parents.\n",
        "    - num_generations (int): Number of generations to run the genetic algorithm.\n",
        "    - best_solution (numpy.ndarray): Best solution found so far.\n",
        "    - best_fitness (float): Fitness score of the best solution.\n",
        "    - population (list): Current population of chromosomes.\n",
        "    - family_pref_matrix (numpy.ndarray): Matrix of family preferences.\n",
        "    - family_sizes (numpy.ndarray): Array of family sizes.\n",
        "    - consolation_gifts (dict): Price Santa pays for each family according to assigned day.\n",
        "    \"\"\"\n",
        "    def __init__(self,\n",
        "                 file_location: str,\n",
        "                 answer_location: str,\n",
        "                 population_size: int = 20,\n",
        "                 mutation_rate: float = 0.9,\n",
        "                 num_generations: int = 1000,\n",
        "                 family_num: int = 5000,\n",
        "                 number_of_days: int = 100,\n",
        "                 min_constraint: int = 125,\n",
        "                 max_constraint: int = 300,\n",
        "                 number_of_preferences: int = 10,\n",
        "                 crossover_rate: float = 0.9):\n",
        "        \"\"\"\n",
        "        Initializes the GeneticScheduler.\n",
        "\n",
        "        Args:\n",
        "            file_location (str): File location of the family preference data.\n",
        "            answer_location (str): File location to save the best solution.\n",
        "            population_size (int): Size of the population in each generation.\n",
        "            mutation_rate (float): Probability to mutate for each chromosome.\n",
        "            num_generations (int): Number of generations to run the genetic algorithm.\n",
        "            family_num (int): Number of families.\n",
        "            number_of_days (int): Number of days Santa's workshop is open.\n",
        "            min_constraint (int): Minimum constraint of people in a workshop per day.\n",
        "            max_constraint (int): Maximum constraint of people in a workshop per day.\n",
        "            number_of_preferences (int): Number of preferences each family provides.\n",
        "            crossover_rate (float): Probability to perform crossover for pair of chromosomes.\n",
        "        \"\"\"\n",
        "        self.file_location = file_location\n",
        "        self.answer_location = answer_location\n",
        "        self.num_days = number_of_days\n",
        "        self.family_num = family_num\n",
        "        self.MIN_PEOPLE = min_constraint\n",
        "        self.MAX_PEOPLE = max_constraint\n",
        "        self.preferences_num = number_of_preferences\n",
        "\n",
        "        self.population_size = population_size\n",
        "        self.mutation_rate = mutation_rate\n",
        "        self.crossover_rate = crossover_rate\n",
        "        self.num_generations = num_generations\n",
        "\n",
        "        # Read csv file with preferences for each family (provided in the task)\n",
        "        family_data = pd.read_csv(self.file_location)\n",
        "        # Retreive column names where the preferences are stored\n",
        "        pref_columns = [f'choice_{i}' for i in range(10)]\n",
        "\n",
        "        # Numpy matrix (5000, 10) with each 10 preferences for each family.\n",
        "        # Note that the size of the matrix can be easily changed for your data.\n",
        "        # However, the Santa's Workshop Tour 2019 provides exactly 5000 families,\n",
        "        # with 10 preferences in each.\n",
        "        self.family_pref_matrix = family_data[pref_columns].to_numpy()\n",
        "\n",
        "        # Numpy array of family sizes, where index is family_id\n",
        "        self.family_sizes = family_data['n_people'].to_numpy()\n",
        "\n",
        "        # Retreive every possible family size in the provided file\n",
        "        unique_family_sizes = np.sort(np.unique(self.family_sizes))\n",
        "\n",
        "        # Price Santa pays for each family according to assigned day\n",
        "        # key: n -- size of the family\n",
        "        # value: list of costs for every possible day assigment (1st preference, etc.)\n",
        "        self.consolation_gifts = {\n",
        "            n: np.array([\n",
        "                0,\n",
        "                50,\n",
        "                50 + 9 * n,\n",
        "                100 + 9 * n,\n",
        "                200 + 9 * n,\n",
        "                200 + 18 * n,\n",
        "                300 + 18 * n,\n",
        "                300 + 36 * n,\n",
        "                400 + 36 * n,\n",
        "                500 + 36 * n + 199 * n,\n",
        "                500 + 36 * n + 398 * n\n",
        "            ])\n",
        "            for n in unique_family_sizes\n",
        "        }\n",
        "\n",
        "        # already during class initialization generate inital population for GA\n",
        "        self.population = self.generate_initial_population()\n",
        "\n",
        "        # Store best found solution ever\n",
        "        self.best_solution = None\n",
        "        # Store best calculated fintess ever\n",
        "        self.best_fitness = float('inf')\n",
        "\n",
        "    def fitness_function(self, chromosome: np.ndarray) -> float:\n",
        "        \"\"\"\n",
        "        Calculates the fitness score of a chromosome.\n",
        "        - It includes restriction on the total number of people attending Santa;\n",
        "        - Preference:cost which is consolation gifts for each family;\n",
        "        - Accounting:penalty which accounts taxes, cleaning costs and etc.\n",
        "        Args:\n",
        "            chromosome (numpy.ndarray): Chromosome representing the assignment of families to days.\n",
        "\n",
        "        Returns:\n",
        "            float: Fitness score of the chromosome.\n",
        "        \"\"\"\n",
        "        # A list of indices corresponding to the days on which families are\n",
        "        # scheduled to visit Santa's workshop according to the chromosome.\n",
        "        assigned_days = [chromosome[i] for i in range(self.family_num)]\n",
        "\n",
        "        fitness = 0\n",
        "\n",
        "        # Counter of total people for every possible day reserved for workshop\n",
        "        day_workshops_number = {d: 0 for d in range(1, self.num_days + 1)}\n",
        "\n",
        "        # Looping over each family;\n",
        "        # fam_size - number of people in family\n",
        "        # day - chosen day for that family in the given chromosome\n",
        "        # preferences - numpy.ndarray of 10 (for this task) preferences of that family\n",
        "        for fam_size, day, preferences in zip(self.family_sizes, assigned_days, self.family_pref_matrix):\n",
        "            # add the number of family members to the daily number of people\n",
        "            day_workshops_number[day] += fam_size\n",
        "\n",
        "            if day not in preferences:\n",
        "                # Calculate the penalty for not assigning a preferred day\n",
        "                fitness += self.consolation_gifts[fam_size][-1]\n",
        "            else:\n",
        "                # Calculate the penalty for assigning a preferred day\n",
        "                index = np.where(preferences == day)[0][0]  # Get the index of the chosen day in preferences\n",
        "                fitness += self.consolation_gifts[fam_size][index]\n",
        "\n",
        "        for n in day_workshops_number.values():\n",
        "            if (n < self.MIN_PEOPLE) or (n > self.MAX_PEOPLE):\n",
        "                # soft constraints for better convergence for the fitness to minimum\n",
        "                # 10**9 is not a magical number, but it were given in the task description\n",
        "                fitness += 10 ** 9\n",
        "\n",
        "        # Calculate accounting:penalty (taxes cost, etc.). The formula were given in the task itself\n",
        "        accounting_penalty = max(0, (day_workshops_number[self.num_days] - 125.0) / 400.0 * day_workshops_number[self.num_days] ** 0.5)\n",
        "        yesterday_workshops_number = day_workshops_number[self.num_days]\n",
        "        for day in range(self.num_days - 1, 0, -1):\n",
        "            today = day_workshops_number[day]\n",
        "            diff = abs(today - yesterday_workshops_number)\n",
        "            accounting_penalty += max(0, (today - 125.0) / 400.0 * today ** (0.5 + diff / 50.0))\n",
        "            yesterday_workshops_number = today\n",
        "\n",
        "        fitness += accounting_penalty\n",
        "        return fitness\n",
        "\n",
        "    def generate_initial_population(self) -> List[np.ndarray]:\n",
        "        \"\"\"\n",
        "        Generates the initial population of chromosomes.\n",
        "\n",
        "        Returns:\n",
        "            list: Initial population of chromosomes.\n",
        "        \"\"\"\n",
        "        population = []\n",
        "        for i in range(self.population_size):\n",
        "            # We are firstly initialize chromosome with zeros\n",
        "            # Later each value will represent chosen day for ith family\n",
        "            # where i is index in the chromosome numpy array\n",
        "            chromosome = np.zeros(self.family_num, dtype=int)\n",
        "\n",
        "            # We'll use intelligent chromosome creation that will try to\n",
        "            # evenly distribute each family by day\n",
        "            days_capacity = {d: 0 for d in range(1, self.num_days + 1)}\n",
        "            # get families indexes as a list\n",
        "            families_order = list(range(self.family_num))\n",
        "            # shuffle this list of family indexes to randomly look at\n",
        "            # families that are not in the any predefined order\n",
        "            random.shuffle(families_order)\n",
        "\n",
        "            for i in families_order:\n",
        "                if chromosome[i] == 0:\n",
        "                    # if chromosome didn't store any day choice yet for that\n",
        "                    # family\n",
        "                    flag = True\n",
        "                    for j in range(self.preferences_num):\n",
        "                        # trying to go through every preference of that family\n",
        "                        day_candidate = self.family_pref_matrix[i][j]\n",
        "                        day_load = days_capacity[day_candidate]\n",
        "                        if day_load + self.family_sizes[i] < (self.MIN_PEOPLE + self.MAX_PEOPLE) / 2:\n",
        "                            # and make sure the day isn't filled with too many families yet.\n",
        "                            # if not, then chose that day and store it in the chromosome\n",
        "                            chromosome[i] = day_candidate\n",
        "                            days_capacity[day_candidate] += self.family_sizes[i]\n",
        "                            flag = False\n",
        "                            # break in order not to consider any other day from preferences\n",
        "                            break\n",
        "                    if flag:\n",
        "                        # if all days from preferences are loaded to much, then just assign\n",
        "                        # random day as a choice for that family in the chromosome\n",
        "                        chromosome[i] = np.random.randint(1, self.num_days + 1)\n",
        "            # append chromosome to the population\n",
        "            population.append(chromosome)\n",
        "\n",
        "        return population\n",
        "\n",
        "    def mutation(self, chromosome: np.ndarray) -> np.ndarray:\n",
        "        \"\"\"\n",
        "        Mutates a chromosome based on mutation rate.\n",
        "\n",
        "        This method introduces random changes into the chromosome to explore new solutions with probability = self.mutation_rate.\n",
        "        It does so by randomly selecting a number of genes to mutate. For each gene selected,\n",
        "        it either swaps the gene's value with one of its preferences or assigns it a random day (type of mutation also based on self.mutation_rate).\n",
        "        This process helps in exploring the solution space more thoroughly and avoiding premature convergence\n",
        "        to suboptimal solutions.\n",
        "\n",
        "        Args:\n",
        "            chromosome (numpy.ndarray): Chromosome to be mutated.\n",
        "\n",
        "        Returns:\n",
        "            numpy.ndarray: Mutated chromosome.\n",
        "        \"\"\"\n",
        "        # Decide whether to mutate the gene based on the mutation rate.\n",
        "        if random.random() < self.mutation_rate:\n",
        "            # Determine the number of mutations to apply based on a random number.\n",
        "            # This introduces variability in the mutation process.\n",
        "            num_mutations = random.randint(1, 5)\n",
        "\n",
        "            # Iterate over the number of mutations to apply.\n",
        "            for _ in range(num_mutations):\n",
        "                # Select a random gene (family) to mutate.\n",
        "                gene_index = random.randint(0, len(chromosome) - 1)\n",
        "\n",
        "                # Decide which type of mutation to apply based on the mutation rate.\n",
        "                if random.random() < self.mutation_rate:\n",
        "                    # Flag to indicate if a suitable mutation has been found.\n",
        "                    flag = True\n",
        "\n",
        "                    # Iterate over the family's preferences.\n",
        "                    for j in range(self.preferences_num):\n",
        "                        # Get the day candidate from the family's preferences\n",
        "                        day_candidate = self.family_pref_matrix[gene_index][j]\n",
        "\n",
        "                        # Calculate the current load of the candidate day.\n",
        "                        day_load = sum(chromosome == day_candidate)\n",
        "\n",
        "                        # Check if the candidate day can accommodate the family\n",
        "                        # without exceeding the maximum capacity.\n",
        "                        if day_load + self.family_sizes[gene_index] < self.MAX_PEOPLE:\n",
        "                            # If the candidate day is suitable, assign the family to\n",
        "                            # this day.\n",
        "                            chromosome[gene_index] = day_candidate\n",
        "\n",
        "                            # Set the flag to False to indicate a successful mutation.\n",
        "                            flag = False\n",
        "\n",
        "                            # Break the loop as the mutation has been applied.\n",
        "                            break\n",
        "\n",
        "                    # If no suitable day was found in the family's preferences,\n",
        "                    # assign the family to a random day.\n",
        "                    if not flag:\n",
        "                        chromosome[gene_index] = np.random.randint(1, self.num_days + 1)\n",
        "                else:\n",
        "                    # If the mutation rate was not met, assign the family to a random day.\n",
        "                    chromosome[gene_index] = np.random.randint(1, self.num_days + 1)\n",
        "\n",
        "        # Return the updated chromosome.\n",
        "        return chromosome\n",
        "\n",
        "    def crossover(self, parent1: np.ndarray, parent2: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:\n",
        "        \"\"\"\n",
        "        Performs crossover between two parents to produce offspring.\n",
        "\n",
        "        This method combines two parent chromosomes to create two new offspring.\n",
        "        The crossover process involves selecting a random number of crossover\n",
        "        points and swapping the segments between these points between the\n",
        "        two parents. This introduces variability in the offspring and helps\n",
        "        in exploring the solution space more thoroughly.\n",
        "\n",
        "        Args:\n",
        "            parent1 (numpy.ndarray): First parent chromosome.\n",
        "            parent2 (numpy.ndarray): Second parent chromosome.\n",
        "\n",
        "        Returns:\n",
        "            tuple: Two offspring chromosomes.\n",
        "        \"\"\"\n",
        "\n",
        "        # Decide whether to perform crossover based on the crossover rate.\n",
        "        if random.random() < self.crossover_rate:\n",
        "          # Randomly choose the number of crossover points from 1 to 3.\n",
        "          # This introduces variability in the crossover process.\n",
        "          num_crossover_points = random.randint(1, 3)\n",
        "\n",
        "          # Generate a sorted list of crossover points.\n",
        "          # This ensures that the crossover points are distinct and ordered.\n",
        "          crossover_points = sorted(random.sample(range(1, len(parent1)), num_crossover_points))\n",
        "\n",
        "          # Add the start and end points to the list of crossover points.\n",
        "          crossover_points.insert(0, 0)\n",
        "          crossover_points.append(len(parent1))\n",
        "\n",
        "          # Initialize the offspring chromosomes.\n",
        "          child1 = []\n",
        "          child2 = []\n",
        "\n",
        "          # Iterate over the crossover points to create the offspring.\n",
        "          for i in range(len(crossover_points) - 1):\n",
        "              # Determine which parent to use for the current segment based on the index.\n",
        "              # This alternates between parents for each segment.\n",
        "              if i % 2 == 0:\n",
        "                  # Use the segment from the first parent.\n",
        "                  segment1 = copy.deepcopy(parent1[crossover_points[i]: crossover_points[i + 1]])\n",
        "                  # Use the segment from the second parent.\n",
        "                  segment2 = copy.deepcopy(parent2[crossover_points[i]: crossover_points[i + 1]])\n",
        "              else:\n",
        "                  # Use the segment from the second parent.\n",
        "                  segment1 = copy.deepcopy(parent2[crossover_points[i]: crossover_points[i + 1]])\n",
        "                  # Use the segment from the first parent.\n",
        "                  segment2 = copy.deepcopy(parent1[crossover_points[i]: crossover_points[i + 1]])\n",
        "\n",
        "              # Append the segments to the corresponding offspring.\n",
        "              child1.extend(segment1)\n",
        "              child2.extend(segment2)\n",
        "\n",
        "          # Convert the lists to numpy arrays for consistency with the rest of the code.\n",
        "          child1 = np.array(child1)\n",
        "          child2 = np.array(child2)\n",
        "\n",
        "          # Return the offspring.\n",
        "          return child1, child2\n",
        "        else:\n",
        "          # If the crossover rate was not met, return the parents unchanged.\n",
        "          return parent1, parent2\n",
        "\n",
        "    def select_parents(self, population: List[np.ndarray], scores: List[float], parents_size: int) -> List[np.ndarray]:\n",
        "        \"\"\"\n",
        "        Selects parents for crossover based on their fitness scores.\n",
        "\n",
        "\n",
        "        This method selects a specified number of parents from the population for crossover.\n",
        "        The selection is based on the fitness scores of the chromosomes, with higher-scoring\n",
        "        chromosomes having a higher chance of being selected. This ensures that the offspring\n",
        "        produced by the crossover process are likely to be better solutions than the parents.\n",
        "\n",
        "        Args:\n",
        "            population (List[numpy.ndarray]): List of chromosomes (parents) from which to select.\n",
        "            scores (List[float]): List of fitness scores corresponding to each chromosome in the population.\n",
        "            parents_size (int): Number of parents to select.\n",
        "\n",
        "        Returns:\n",
        "            List[numpy.ndarray]: Selected parents.\n",
        "        \"\"\"\n",
        "        # Initialize an empty list to store the selected parents.\n",
        "        selected_parents = []\n",
        "\n",
        "        # Iterate over the number of parents to select.\n",
        "        for _ in range(parents_size):\n",
        "            # Randomly select two indices from the population.\n",
        "            idx1, idx2 = np.random.choice(range(len(population)), size=2, replace=False)\n",
        "\n",
        "            # Compare the fitness scores of the two selected chromosomes.\n",
        "            # The chromosome with the lower score is more likely to be selected,\n",
        "            # as it represents a better solution.\n",
        "            if scores[idx1] < scores[idx2]:\n",
        "                # If the first chromosome has a lower score, add it to the selected parents.\n",
        "                selected_parents.append(copy.deepcopy(population[idx1]))\n",
        "            else:\n",
        "                # If the second chromosome has a lower score, add it to the selected parents.\n",
        "                selected_parents.append(copy.deepcopy(population[idx2]))\n",
        "\n",
        "        # Return the list of selected parents.\n",
        "        return selected_parents\n",
        "\n",
        "    def replace(self, offsprings: List[np.ndarray], population_scores: List[float]) -> List[np.ndarray]:\n",
        "        \"\"\"\n",
        "        Replaces the old population with the new one (including offspring).\n",
        "\n",
        "        This method combines the current population with the newly produced offspring and selects the best\n",
        "        chromosomes to form the new population. The selection is based on the fitness scores of the chromosomes,\n",
        "        with higher-scoring chromosomes having a higher chance of being selected. This ensures that the new\n",
        "        population is likely to be better than the old one.\n",
        "\n",
        "\n",
        "        Args:\n",
        "            offsprings (List[numpy.ndarray]): List of offspring chromosomes produced by\n",
        "                                              crossover and mutation.\n",
        "            population_scores (List[float]): List of fitness scores corresponding to\n",
        "                                             each chromosome in the current population.\n",
        "\n",
        "        Returns:\n",
        "            List[numpy.ndarray]: New population after replacement.\n",
        "        \"\"\"\n",
        "        # Combine the current population with the offsprings.\n",
        "        # Each offspring is paired with its fitness score, which is calculated by applying the fitness function.\n",
        "        combined_population = [(chromosome, score) for chromosome, score in zip(self.population, population_scores)] + \\\n",
        "                            [(offsprings[i], self.fitness_function(offsprings[i])) for i in range(len(offsprings))]\n",
        "\n",
        "        # Sort the combined population based on the fitness scores.\n",
        "        # This ensures that the chromosomes with the lowest scores are at the beginning of the list.\n",
        "        combined_population.sort(key=lambda x: x[1])\n",
        "\n",
        "        # Select the best chromosomes from the combined population to form the new population.\n",
        "        # The size of the new population is the same as the original population size.\n",
        "        new_population = [chromosome for chromosome, score in combined_population[:self.population_size]]\n",
        "\n",
        "        # Return the new population.\n",
        "        return new_population\n",
        "\n",
        "    def save_solution(self):\n",
        "        \"\"\"\n",
        "        Saves the best solution to a CSV file.\n",
        "\n",
        "        This method retrieves the best chromosome found by the genetic algorithm and saves it to a CSV file.\n",
        "        The best chromosome represents the optimal assignment of families to days, based on the fitness function.\n",
        "        The solution is saved in a format that includes the family ID and the assigned day for each family.\n",
        "\n",
        "        \"\"\"\n",
        "        # Retrieve the best chromosome - the chromosome with the lowest fitness score,\n",
        "        # representing the best solution found.\n",
        "        best_chromosome = self.best_solution\n",
        "\n",
        "        # Initialize an empty list to store the solution.\n",
        "        # Each element in the list will be a list containing the family ID and the assigned day.\n",
        "        answer = []\n",
        "        for family_id, day in enumerate(best_chromosome):\n",
        "            # Append the family ID and the assigned day to the solution list.\n",
        "            answer.append([family_id, day])\n",
        "\n",
        "        # Convert the solution list to a pandas DataFrame.\n",
        "        # The DataFrame has two columns: 'family_id' and 'assigned_day'.\n",
        "        submission_df = pd.DataFrame(answer, columns=['family_id', 'assigned_day'])\n",
        "\n",
        "        # Save the DataFrame to the specified file path\n",
        "        # This file will contain the best solution found by the genetic algorithm.\n",
        "        submission_df.to_csv(self.answer_location, index=False)\n",
        "\n",
        "        # Print a message indicating that the solution was successfully saved.\n",
        "        print(\"Solution was successfully saved to\", self.answer_location)\n",
        "\n",
        "    def run(self):\n",
        "        \"\"\"\n",
        "        Runs the genetic algorithm for scheduling.\n",
        "\n",
        "        This method executes the genetic algorithm for a specified number of generations. It includes the\n",
        "        initialization of the population, the evaluation of fitness for each chromosome, the selection of\n",
        "        parents for crossover, the crossover and mutation of offspring, and the replacement of the old\n",
        "        population with the new one. The algorithm also includes a mechanism to regenerate the initial\n",
        "        population if the fitness does not improve for a certain number of generations, to avoid\n",
        "        premature convergence to suboptimal solutions.\n",
        "        \"\"\"\n",
        "        log_file = open('logs.txt', 'w')\n",
        "        # Iterate over the specified number of generations.\n",
        "        for generation in range(self.num_generations):\n",
        "            # Evaluate the fitness of each chromosome in the population\n",
        "            population_scores = [self.fitness_function(chromosome) for chromosome in self.population]\n",
        "\n",
        "            # Remember the current best fitness\n",
        "            current_best_fitness = min(population_scores)\n",
        "            # If the current best fitness is lower than the previous best, update the best solution and fitness.\n",
        "            if current_best_fitness < self.best_fitness:\n",
        "                self.best_solution = copy.deepcopy(self.population[population_scores.index(current_best_fitness)])\n",
        "                self.best_fitness = current_best_fitness\n",
        "\n",
        "            # Every 200 generations, save the current best solution.\n",
        "            if generation % 200 == 0:\n",
        "              self.save_solution()\n",
        "\n",
        "            # Print the current generation number, the best fitness, the best fitness in the current population,\n",
        "            # and the mean fitness of the population.\n",
        "            print(f\"Generation {generation + 1}: Best fitness = {self.best_fitness}\")\n",
        "            log_file.write(f\"Generation {generation + 1}: Best fitness = {self.best_fitness}\\n\")\n",
        "\n",
        "            # Select parents for crossover\n",
        "            parents = self.select_parents(self.population, population_scores, self.population_size // 2)\n",
        "\n",
        "            # Perform crossover to create offsprings\n",
        "            offsprings = []\n",
        "            for i in range(len(parents) - 1):\n",
        "                child1, child2 = self.crossover(parents[i], parents[i + 1])\n",
        "                offsprings.append(child1)\n",
        "                offsprings.append(child2)\n",
        "\n",
        "            # Apply mutation to the offspring\n",
        "            for i in range(len(offsprings)):\n",
        "                offsprings[i] = self.mutation(offsprings[i])\n",
        "\n",
        "            # Replace the old population with the new one (including offspring)\n",
        "            self.population = self.replace(offsprings, population_scores)\n",
        "\n"
      ],
      "metadata": {
        "id": "oULLjoaoaLuq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "data_file_path = 'family_data.csv'\n",
        "answer_file_path = 'submission.csv'\n",
        "scheduler = GeneticScheduler(data_file_path, answer_file_path)  # Initialize the scheduler\n",
        "scheduler.run()  # Execute the genetic algorithm\n"
      ],
      "metadata": {
        "id": "3zoL9vTI4STX"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}